src/core/data/transform.h

```diff
@@ -51,6 +52,13 @@ struct TransformStack : public Transform {
  public:
   void dump() const;
 
+ public:
+  void return_promoted_dims(std::vector<int32_t>& dims)
+  {
+    if (nullptr != transform_) transform_->return_promoted_dims(dims);
+    if (nullptr != parent_) parent_->return_promoted_dims(dims);
```

Comment:
You probably need to process the parent before the current node in the stack, see comments in `transform.cc`.